lib/msf/core/post/windows/mssql.rb

```diff
@@ -109,13 +109,27 @@ def check_sqlcmd
         # @param [String] query the query to execute
         # @param [String] instance the SQL instance to target
         # @param [String] server the SQL server to target
+        # @param [String] database the database to connect to upon successfully authenticating
+        # @param [String] username the username to authenticate as
+        # @param [String] password the password to authenticate with
         # @return [String] the result of query
-        def run_sql(query, instance = nil, server = '.')
+        def run_sql(query, instance = nil, server = '.', database: nil, username: nil, password: nil)
           target = server
           if instance && instance.downcase != 'mssqlserver'
             target = "#{server}\\#{instance}"
           end
-          cmd = "#{@sql_client} -E -S #{target} -Q \"#{query}\" -h-1 -w 200"
+          cmd = "#{@sql_client}"
+          cmd += " -d #{database}" unless database.nil? || database.empty?
+          if username && password
+            if username.empty?
+              print_error("Username field was blank!")
+              return ""
+            end
+            cmd += " -U \"#{username}\" -P \"#{password}\""
```

Comment:
Do we need to do anything special if `password` contains characters such as quotes or `\` etc?

Reply:
Good point, I would imagine this might be a concern?

Reply:
FWIW, that's kind of a wide spread problem. You can see the same pattern used in alot of the Unix post API commands where the argument is quoted, but not escaped.

Reply:
Could probably solve this with a `.gsub!` call. I'll see about creating a solution for this.

Reply:
After further discussion within the team it was determined that the backslash solution we currently have might not be the best option as Windows character escaping is a complex task and we should likely develop a library for this. See https://github.com/rapid7/metasploit-framework/issues/14742#issuecomment-778210006 where this issue was discussed in more detail along with potential solutions.

Reply:
Fixed things up with https://github.com/rapid7/metasploit-framework/pull/16948/commits/4c9ef22c59bd0aac7c175597d6f9d511d7a6cc8e, should address these concerns.